design/specs/visual-query-builder.md

```diff
@@ -0,0 +1,140 @@
+---
+title: Visual Query Builder
+description: 
+published: true
+date: 2022-03-08 18:02:53
+tags: 
+editor: markdown
+dateCreated: 2022-03-08 18:02:58
+---
+
+## Context
+
+This document specifies the design for a visual query builder named `Data Explorer` within the Mathesar application. Through `Data Explorer`, users will be able to analyze data across one or multiple tables of a schema and save the resulting tables as views.
+
+Additionally, users will be able to open existing views in `Data Explorer`, as long as its query commands and functions are supported.
+
+## Scenarios
+
+## 1. Selecting the Base Table
+
+The base table selection is the first step when starting a new query, and it will determine which columns and links are available for selection.
+
+The base table also determines the automatic joins that are performed when the user adds columns from linked tables.
+
+### Changing the base table
+
+Once the input table has been created, changing the base table will result in the loss of all progress.
+
+### 1.1 Base Table Options
+
+The base table selector will display a list of all available tables in the current schema. Initially, this list will not include views which will be added in later iterations.
+
+To select a table, the user will find it on the list and click on the desired table option.
+
+Selecting the base table will enable the interface controls for adding additional steps, which were initially disabled. No action is available before a base table is selected.
+
+Once a selection has been made, the user will be prompted to proceed with the input table configuration, starting with column selection.
+
+[BASE TABLE SELECT WIREFRAME]
+
+## 2. Creating the Input Table
+
+The input table is created by selecting columns from the base table or creating new ones via formulas.
+
+Various options are available for input columns, including filters or aggregations that can help users control which data they want to show and how.
```

Comment:
Filters and aggregations are only available for columns that have multiple records.

For example, if you have the following structure:

---
### Movie

| ID | Title | Runtime | Year |
|-|-|-|-|
| 1 | Encanto | 109 | 2021 |

### Actor

| ID | Name |
|-|-|
| 462 | Stephanie Beatriz |
| 463 | John Leguizamo |

### Movie Actor Map

| ID | Actor ID | Movie ID | Character |
|-|-|-|-|
| 34 | 462 | 1 | Mirabel Madrigal |
| 35 | 463 | 1 | Bruno Madrigal |

---

Imagine the user sets the base table to "Movie" and adds "Movie > Title" as the first column. 

Now if they add "Movie > Runtime", there won't be any filters or aggregations available for that column because each movie has only one runtime, there's nothing to filter or aggregate.

Instead, if they add "Actor > Name", there will be filters and aggregations available since there are multiple actors associated with a single movie.